docs/RFCS/gateway_batch.md

```diff
@@ -0,0 +1,147 @@
+worry about, and so from that perspective, splitting the `Batch` is
+`TxnCoordSender`'s job.
+
+But it's desirable to keep the actual "cutting" of the batch in `DistSender`
+since that's where the range descriptors are cached and looked up. It's also
+desirable to have the cutting and sending in one context, so that all the
+information is available to deal with errors (and specifically cache
+invalidation) appropriately.
+
+I think the snippet interface is a good starting point for discussion:
+```go
+// SendBatch is called by (*TxnCoordSender).Send().
+func (ds *DistSender) SendBatch(b *proto.BatchRequest, preHook func([]*proto.BatchRequest)) proto.BatchResponse {
+  // Retry loop here.
+  // ...
+
+    // Chop up the batch. This is already in a retry loop.
+    var chopped []*proto.BatchRequest = ds.Split(b)
+    // Execute the callback, which may rearrange some items and update
+    // `EndTransaction`.
+    // preHook **must** be idempotent! It may be executed multiple times if
+    // stale descriptors causes a retryable error.
+    b = preHook(b)
+
+    // Send the reordered batch requests one by one, much like ds.Send()
+    // ...
+
+  // Recombine the multi-range requests back into one.
+  // ...
+
+  return response
+}
+```
+
+That would cover the above bases: `TxnCoordSender` gets to inspect the chunks of
+the batch and reorders them such that `EndTransaction` comes last; it also makes
+sure that `EndTransaction` has the complete list of intents. Another reasonable
+optimization, namely sending chunks of the batch in parallel, could be carried
+out similarly: In this case, `preHook` would not only operate on the slice, but
+return a `Plan` consisting of different stages, each containing multiple chunks
+(extremes: each stage consisting of one one-element batch = sequential querying
+as we do it now; one stage of `N` batches (to different ranges) = completely
+parallel). Non-transactions would go completely parallel, transactions are a
+bit more involved (going parallel might lay down a lot of intents
+and do a lot of work for nothing) but will always have a last stage with the
+chunk containing `EndTransaction`.
+
+## Placeholders and Multi-Range requests
+
+We need to introduce a placeholder entry in a `Batch` (`nil`?) so that the
+chunks of a batch can be merged.
+Also, clients rely on the order in a `Batch` being fixed.
+
+Some result types (notably `Scan`) implement a `Combinable` interface which
+is used to glue together the subranges of the result contained on different
+`Range`s. Combining those is easiest if the index of request in a `Batch` is
+equal across all chunks. For any index with at least two non-placeholders requests,
+these requests must implement `Combinable`. Similar considerations hold for
+truncation, which is more complicated: once over limit, all future chunks of
+the batch should turn that request into a placeholder to avoid querying
+unnecessary further data.
```

Comment:
I'm assuming that `DistSender` is going to be responsible for managing this chunk structure, which is complex behavior which requires special handling depending on the specific operations being executed.

With that in mind, I'm confused as to why `DistSender` can't be specially aware of `EndTransaction` when batching. That seems no more complex than knowing that certain operations have a `Combinable` result type, and it would obviate the need for the `preHook` argument outlined above.